src/basic/chase.c

```diff
@@ -532,19 +551,27 @@ int chase(const char *path, const char *original_root, ChaseFlags flags, char **
 
         if (ret_path) {
                 if (!FLAGS_SET(flags, CHASE_EXTRACT_FILENAME)) {
-                        _cleanup_free_ char *q = NULL;
 
-                        q = path_join(empty_to_root(root), p);
-                        if (!q)
-                                return -ENOMEM;
+                        /* When "root" points to the root directory, the result of chaseat() is always
+                         * absolute, hence it is not necessary to prefix with the root. When "root" points to
+                         * a non-root directory, the result path is always normalized and relative, hence
+                         * we can simply call path_join() and not necessary to call path_simplify().
+                         * Note that the result of chaseat() may start with "." (more specifically, it may be
+                         * "." or "./"), and we need to drop "." in that case. */
+
+                        if (empty_or_root(root))
+                                assert(path_is_absolute(p));
```

Comment:
I've just run into the following assert when someone I helped mounted two zfs datasets at /. Not sure if this falls under bug or wrong usage.

Reply:
Could you, please, file a new issue with some more details? So we can look into what's going on.

Reply:
#29559 I hope that's not to vague 😅